projector-html-purs/src/Projector/Html/Backend/Purescript/Rewrite.hs

```diff
@@ -0,0 +1,236 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+module Projector.Html.Backend.Purescript.Rewrite (
+    rewriteModule
+  , rewriteExpr
+  , rules
+  ) where
+
+
+import qualified Control.Monad.Trans.State as State
+
+import qualified Data.Text as T
+
+import           P
+
+import           Projector.Core
+import qualified Projector.Core as Core
+import qualified Projector.Html.Backend.Rewrite as Rewrite
+import qualified Projector.Html.Core.Library as CL
+import           Projector.Html.Data.Module
+import           Projector.Html.Data.Prim
+
+
+rewriteModule :: ModuleName -> Module HtmlType PrimT a -> (ModuleName, Module HtmlType PrimT a)
+rewriteModule mn (Module tys imports exprs) =
+  let exprs' = fmap (\(ModuleExpr ty e) -> ModuleExpr ty (rewriteExpr (Just mn) e)) exprs
+  in (mn, Module tys imports exprs')
+
+rewriteExpr :: Maybe ModuleName -> Expr PrimT a -> Expr PrimT a
+rewriteExpr mn =
+  Core.rewriteFix (rules mn) . Core.rewriteFix Rewrite.globalRules
+
+-- * Drop all comments, they don't make sense in virtual-dom world
+-- * Replace constructors with DOMLike typeclass methods
+rules :: Maybe ModuleName -> [RewriteRule PrimT a]
+rules mmn =
+  fmap Rewrite [
+      -- Replace HTML model with DOMLike functions.
+      -- These rules are important for correctness - won't work without these.
+      (\case ECon a (Constructor "Plain") _ [x] ->
```

Comment:
Maybe a dumb question, any way to share some of this stuff with the haskell version, at least for the stringy parts?

Reply:
some of the optimising rewrites could probably be shared, just not sure where to put the code.